### PR TITLE
Added #14979: add checkout to location and assets functionality to accessories

### PR DESCRIPTION
**Pull request by arne-kroeger**
Added #14979: add checkout functionality to accessories

# Description

Currently it is just possible to checkout accessories to a user and not to a location or asset like it is possible for assets. This would cover many use cases where a workplace definition can be done with a location or an assignment of accessories to a room. There are several uses case descriptions in the linked issues (see below). 

Added #14665 
Added #12381 
Added #12452 
Added #11820
Added #10956 
Added #5140  (parts of it)
Added #14979 (parts of it)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tests\Feature\Checkouts\Ui\AccessoryCheckoutTest::testAccessoryCanBeCheckedOutToLocationWithQuantity
- [x] Tests\Feature\Checkouts\Ui\AccessoryCheckoutTest::testAccessoryCanBeCheckedOutToAssetWithQuantity

**Test Configuration**:
* PHP version: 8.3.9
* MariaDB version - 10.11
* Webserver version - php server
* OS version - Mac OS 

# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by snipe** [comment]
We haven't built this functionality because we don't really want it. Our position has always been that you should be checking items out to users, since an asset or location cannot be held responsible for missing items.

File: app/Http/Controllers/Api/AccessoriesController.php [hunk 1]
```diff
@@ -197,28 +200,23 @@ public function checkedout($id, Request $request)
...
 
         if ($total < $offset) {
             $offset = 0;
         }
 
-        $accessory_users = $accessory->users()->skip($offset)->take($limit)->get();
+        $accessory_checkouts = $accessory->checkouts()->skip($offset)->take($limit)->get();
 
         if ($request->filled('search')) {
-            $accessory_users = $accessory->users()
```
Comment (snipe) [inline app/Http/Controllers/Api/AccessoriesController.php:210]: Why was this removed? This allows searching on user's names in the checkout list.

Comment (snipe) [reply app/Http/Controllers/Api/AccessoriesController.php:210]: I see, you added a query scope further down the line.

**Comment by snipe** [comment]
Let me spend some time looking at this and we'll consider whether we want to take this. As I mentioned, we didn't really want this functionality since it goes against the principles of what an asset system is supposed to do, but I'll consider it.

File: app/Http/Controllers/Api/AccessoriesController.php (hunk 1 above)
Comment (arne-kroeger) [reply app/Http/Controllers/Api/AccessoriesController.php:210]: This has been replaced with a search over all three kinds of models depending of the checkout type (see advancedTextSearch of the new AccessoryCheckout model).

**Comment by arne-kroeger** [comment]
> We haven't built this functionality because we don't really want it. Our position has always been that you should be checking items out to users, since an asset or location cannot be held responsible for missing items.

As there is an option to define a manager for a location I see a good way to assign a responsiblity here. 

Our use case is also important to desk sharing topics where some of the accessories are part of a bookable working place for the employees. To give these items out on a daily basis is a lot of work. The idea is to assign locations via an additional workplace booking system on a time based ticket and handle the responsibility by it.

**Comment by snipe** [comment]
> As there is an option to define a manager for a location I see a good way to assign a responsibility here.

Sure, but we don't have anything built in for them to be able to accept, etc. Right now it's just a field for reference, there are no controls there, and very few people use managers for locations anyway.

**Comment by snipe** [comment]
>  To give these items out on a daily basis is a lot of work. The idea is to assign locations via an additional workplace booking system on a time based ticket and handle the responsibility by it.

I get that, but it opens the doors to people in other situations who don't use an external "responsible for" application. 

As I said, let me think about it.

File: database/migrations/2024_07_26_143301_add_checkout_for_all_types_to_accessories.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        DB::update('update '.DB::getTablePrefix().'accessories_checkout set assigned_type = \'App\\\\Models\\\\User\' where assigned_type is null', []);
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        if (Schema::hasTable('accessories_checkout')) {
+            Schema::rename('accessories_checkout', 'accessories_users');
```
Comment (snipe) [inline database/migrations/2024_07_26_143301_add_checkout_for_all_types_to_accessories.php:40]: We probably want to drop the column created in the up migration as well, no?

Comment (arne-kroeger) [reply database/migrations/2024_07_26_143301_add_checkout_for_all_types_to_accessories.php:40]: Yes, I will adjust that.

File: app/Http/Controllers/Api/AccessoriesController.php [hunk 3]
```diff
@@ -316,19 +314,19 @@ public function checkout(AccessoryCheckoutRequest $request, Accessory $accessory
      */
     public function checkin(Request $request, $accessoryUserId = null)
     {
-        if (is_null($accessory_user = DB::table('accessories_users')->find($accessoryUserId))) {
+        if (is_null($accessory_checkout = DB::table('accessories_checkout')->find($accessoryUserId))) {
```
Comment (uberbrady) [inline app/Http/Controllers/Api/AccessoriesController.php]: Since we're already modifying this line, is there maybe a way for us to *not* say `DB::table()` here? That always feels to me like a code smell. And the line's already being altered, is there a way to do this properly polymorphic and make it a little cleaner?

File: app/Http/Controllers/Api/AccessoriesController.php [hunk 4]
```diff
@@ -316,19 +314,19 @@ public function checkout(AccessoryCheckoutRequest $request, Accessory $accessory
...
         $this->authorize('checkin', $accessory);
 
-        $logaction = $accessory->logCheckin(User::find($accessory_user->assigned_to), $request->input('note'));
+        $logaction = $accessory->logCheckin(User::find($accessory_checkout->assigned_to), $request->input('note'));
 
         // Was the accessory updated?
-        if (DB::table('accessories_users')->where('id', '=', $accessory_user->id)->delete()) {
-            if (! is_null($accessory_user->assigned_to)) {
-                $user = User::find($accessory_user->assigned_to);
+        if (DB::table('accessories_checkout')->where('id', '=', $accessory_checkout->id)->delete()) {
```
Comment (uberbrady) [inline app/Http/Controllers/Api/AccessoriesController.php]: Another DB::table here - possibly a way to not do this? Not sure....

File: app/Models/Accessory.php [hunk 5]
```diff
@@ -253,9 +253,10 @@ public function getImageUrl()
      * @since [v3.0]
      * @return \Illuminate\Database\Eloquent\Relations\Relation
      */
-    public function users()
+    public function checkouts()
     {
-        return $this->belongsToMany(\App\Models\User::class, 'accessories_users', 'accessory_id', 'assigned_to')->withPivot('id', 'created_at', 'note')->withTrashed();
+        return $this->hasMany(\App\Models\AccessoryCheckout::class, 'accessory_id')
```
Comment (uberbrady) [inline app/Models/Accessory.php:258]: EXQUISITE!

**Review by uberbrady — COMMENTED** [review]
This is some genuine good-looking code - I have a few small nits I've pointed out, but I feel like 80% of this is just handling that the table name has changed. This really does look pretty good to me. Wonderful, very well-written, tight, careful code. Thank you so much for your contribution!

File: app/Http/Controllers/Api/AccessoriesController.php (hunk 3 above)
Comment (arne-kroeger) [reply app/Http/Controllers/Api/AccessoriesController.php]: Adjusted

File: app/Http/Controllers/Api/AccessoriesController.php (hunk 4 above)
Comment (arne-kroeger) [reply app/Http/Controllers/Api/AccessoriesController.php]: Adjusted

File: database/migrations/2024_07_26_143301_add_checkout_for_all_types_to_accessories.php (hunk 2 above)
Comment (arne-kroeger) [reply database/migrations/2024_07_26_143301_add_checkout_for_all_types_to_accessories.php:40]: Adjusted

**Comment by snipe** [comment]
@arne-kroeger I'm seeing some failing tests here - do you think you can take a look at this. I'm reluctantly inclined to take this (not because your code is bad - it's great - just from a product perspective), but we do need the tests to pass.

**Comment by snipe** [comment]
I think you might have missed a bit in the seeder, but I think I fixed it

https://github.com/snipe/snipe-it/blob/39727820330d4abe434b1491ffe55fc1a6308146/database/seeders/AccessorySeeder.php#L19

**Comment by snipe** [comment]
Something still seems off here. On develop.snipeitapp.com (the live dev branch) I just checked out an accessory to a location, and now the accessories page loads with an empty table, and the location does not show any accessories assigned.

**Comment by snipe** [comment]
False alarm on accessories not loading (weirdly, it was a 429 error, which probably bears investigation), but the location doesn't show anything checked out to it.

Screenshot 2024-07-29 at 8 36 10 PM
Screenshot 2024-07-29 at 8 36 20 PM

**Comment by arne-kroeger** [comment]
> False alarm on accessories not loading (weirdly, it was a 429 error, which probably bears investigation), but the location doesn't show anything checked out to it.

Thats a good point.

Currently the location would show Accessories which are placed there (in terms of the field location inside the accessory). What would be you're recommendation? 

A mixed list of placed and checked out accessories?

**Comment by snipe** [comment]
I think you're starting to see why we didn't build this. It's always more complicated than it seems. I suppose we should handle it the way we do locations, where there are assets that belong to that location, and then there are assets that are checked out to that location. Different tabs. I don't think there's a way to make it not-confusing. 

Screenshot 2024-07-30 at 11 38 50 AM

**Comment by arne-kroeger** [comment]
@snipe - Yeah, but I think in the end it will be a good enhancement. I will prepare an integration for it and prepare it.

**Comment by snipe** [comment]
This also misses the "Print all assigned" views on locations, etc. We need to pull accessories into there as well.

**Comment by snipe** [comment]
This may also require a rework of the locations layout altogether, as you can see adding additional tabs here causes them to wrap, and this could potentially only get much worse with other languages that use longer words.

Screenshot 2024-08-03 at 12 24 25 PM

I may have to revert this PR until it's more complete and thought through. :( Right now, letting people check things out to a location but not actually seeing them on the locations page will be a very confusing UI.

File: app/Http/Controllers/Users/BulkUsersController.php [hunk 6]
```diff
@@ -219,7 +219,7 @@ public function destroy(Request $request)
 
         $users = User::whereIn('id', $user_raw_array)->get();
         $assets = Asset::whereIn('assigned_to', $user_raw_array)->where('assigned_type', \App\Models\User::class)->get();
-        $accessories = DB::table('accessories_users')->whereIn('assigned_to', $user_raw_array)->get();
+        $accessories = DB::table('accessories_checkout')->whereIn('assigned_to', $user_raw_array)->get();
```
Comment (snipe) [inline app/Http/Controllers/Users/BulkUsersController.php:222]: I think this query will also not work, since it's not constraining to users. Location ID could get pulled from that list as well as User ID 4.
